### PR TITLE
Extending Symfony form with upload image field tutorial

### PR DESCRIPTION
**Pull request by tdavidsonas88**
In this tutorial we are going to build module which extends Suppliers form with the Upload image field.

Partially solves: https://github.com/PrestaShop/PrestaShop/issues/17138

### CODE REVIEW

**Comment by matks** [comment]
@tdavidsonas88 Time to finish this PR then 😊

**Comment by tdavidsonas88** [comment]
> @tdavidsonas88 Time to finish this PR then

It's really a good time to finish it for me. :)

File: src/content/1.7/modules/sample-modules/grid-and-identifiable-object-form-hooks-usage.md [hunk 1]
```diff
@@ -314,3 +314,144 @@ private function updateCustomerReviewStatus(array $params)
 when we created the switch type form we named it `is_allowed_for_review`. By using the same name we can get the state (on or off).
 This hook receives from `$params` the form data, that you can retrieve like this: `$params['form_data']`.
 All the form data is available here, including `is_allowed_for_review` data which comes from the switch.
+
+The whole main class could look similar to:
```
Comment (matks) [inline src/content/1.7/modules/sample-modules/grid-and-identifiable-object-form-hooks-usage.md:318]: Is it the right place to put this statement ? The text before seems unrelated to your added statement. Maybe we need a transition between them ?

File: src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md [hunk 2]
```diff
@@ -0,0 +1,586 @@
...
+(SELL -> Catalog -> Brands & Suppliers). You will learn how to create:
+
+- Main module class
+- Installer class
+- Create Symfony controller
+- Doctrine entity
+- Repository class
+- Image Uploader class
+- Twig View template
+
```
Comment (matks) [inline src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md:24]: I suggest we explain quickly what each class will do, that gives context to the reader.
```
- Main module class
- Installer class
- Create Symfony controller
- Doctrine entity
- Repository class
- Image Uploader class
- Twig View template
```
⬇️ 
```
- Main module class: main module entry point and hook entry point
- Installer class: responsible for module installation and uninstallation process
- Create Symfony controller: needed as we add a new "delete image" controller action
- Doctrine entity: this model is responsible for the image data persistence
- Repository class: this model is for image database search and retrieve
- Image Uploader class: this class is responsible for image upload process
- Twig View template: needed for display
```

File: src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md [hunk 3]
```diff
@@ -0,0 +1,586 @@
...
+```php
+<?php
+declare(strict_types=1);
+
+// use statements
+
+if (!defined('_PS_VERSION_')) {
+    exit;
+}
+
```
Comment (matks) [inline src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md:38]:
```suggestion
// needed as use Composer to autoload this module
```

File: src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md [hunk 4]
```diff
@@ -0,0 +1,586 @@
...
+                return false;
+            }
+        }
+
+        return true;
+    }
+}
+```
+
+Let's use `Installer` class inside the main module class by adding code snippet
```
Comment (matks) [inline src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md:183]: Nice 👍 step by step

File: src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md [hunk 5]
```diff
@@ -0,0 +1,586 @@
...
+     */
+    public function uninstall()
+    {
+        $installer = new Installer();
+
+        return $installer->uninstall() && parent::uninstall();
+    }
+```
+
+Let's create `SupplierExtraImage` entity class:
```
Comment (matks) [inline src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md:212]:
```suggestion
Let's create `SupplierExtraImage` entity class. We use [Doctrine](https://devdocs.prestashop.com/1.7/modules/concepts/doctrine/) which is available for PrestaShop modules since version 1.7.6.
```

File: src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md [hunk 6]
```diff
@@ -0,0 +1,586 @@
...
+- Image Uploader class
+- Twig View template
+
+### Main module class
+
+Let's create main module class `DemoExtendSymfonyForm2`
+
+```php
+<?php
+declare(strict_types=1);
```
Comment (matks) [inline src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md:31]:
```suggestion
// since this module is compatible with PS 1.7.7 and later, we can use PHP7 strict types because PHP5 support has been dropped for PS 1.7.7
declare(strict_types=1);
```

File: src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md [hunk 7]
```diff
@@ -0,0 +1,586 @@
...
+        $em = $this->getEntityManager();
+        if ($supplierExtraImage) {
+            $em->remove($supplierExtraImage);
+            $em->flush();
+        }
+    }
+}
+```
+
+Let's create hook `hookActionSupplierFormBuilderModifier` function inside Main module class:
```
Comment (matks) [inline src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md:350]:
```suggestion
Let's create hook `hookActionSupplierFormBuilderModifier` function inside Main module class. This is a hook available for [CRUD forms](https://devdocs.prestashop.com/1.7/modules/sample-modules/grid-and-identifiable-object-form-hooks-usage/) in PrestaShop Symfony pages.
```

File: src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md [hunk 8]
```diff
@@ -0,0 +1,586 @@
...
+    {
+        /** @var SupplierExtraImageRepository $supplierExtraImageRepository */
+        $supplierExtraImageRepository = $this->get(
+            'prestashop.module.demoextendsymfonyform.repository.supplier_extra_image_repository'
+        );
+
+        $translator = $this->getTranslator();
+        /** @var FormBuilderInterface $formBuilder */
+        $formBuilder = $params['form_builder'];
+        $formBuilder
```
Comment (matks) [inline src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md:366]:
```suggestion
        // we add to the Symfony form a 'upload_image_file' that will be used by BO user to upload image files
        $formBuilder
```

File: src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md [hunk 9]
```diff
@@ -0,0 +1,586 @@
...
+            ->add('upload_image_file', FileType::class, [
+                'label' => $translator->trans('Upload image file', [], 'Modules.DemoExtendSymfonyForm'),
+                'required' => false,
+            ]);
+
+        /** @var SupplierExtraImage $supplierExtraImage */
+        $supplierExtraImage = $supplierExtraImageRepository->findOneBy(['supplierId' => $params['id']]);
+        if ($supplierExtraImage && file_exists(_PS_SUPP_IMG_DIR_ . $supplierExtraImage->getImageName())) {
+            $formBuilder
```
Comment (matks) [inline src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md:375]:
```suggestion
            // when an image is already registered for this supplier, we add to the Symfony form a
            // 'image_file' to provide a preview input to BO user and also provide a "delete" button
            $formBuilder
```

File: src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md [hunk 10]
```diff
@@ -0,0 +1,586 @@
...
+                ),
+                UploadedImageConstraintException::UNRECOGNIZED_FORMAT
+            );
+        }
+    }
+
+}
+```
+
+Let's create hook `hookActionAfterUpdateSupplierFormHandler` inside main module class:
```
Comment (matks) [inline src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md:534]: :+1 Nice and efficient

File: src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md [hunk 11]
```diff
@@ -0,0 +1,586 @@
...
+---
+
+# Extending Symfony form with upload image field
+{{< minver v="1.7.7" title="true" >}}
+
+
+## Introduction
+
+In this tutorial we are going to build a module which extends `Suppliers` form 
+(SELL -> Catalog -> Brands & Suppliers). You will learn how to create:
```
Comment (matks) [inline src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md:15]:
```suggestion
(SELL -> Catalog -> Brands & Suppliers). This module will address the following need "I'd like to add a 'upload image' field to the Supplier Add/Edit form because I want to display a new logo for each supplier on my shop. So this new field must allow me to add, edit and delete image files linked to a supplier. Obviously I expect the uploaded files to be stored in a consistent way with how PrestaShop usually stores such files."

We are going to create the module to address this need by using PrestaShop hooks in the Add/Edit Supplier Back-Office page and we will follow some software development best practices such as SOLID principles to make our code as clear and maintainable as possible.
You will learn how to create:
```

I'm not sure about the "we will follow some software development best practices such as SOLID principles to make our code as clear and maintainable as possible." part, what is your opinion @tdavidsonas88 ?

File: src/content/1.7/modules/sample-modules/grid-and-identifiable-object-form-hooks-usage.md (hunk 1 above)
Comment (tdavidsonas88) [reply src/content/1.7/modules/sample-modules/grid-and-identifiable-object-form-hooks-usage.md:318]: Maybe it's better to provide a link to example module instead of putting whole main class here?

Comment (tdavidsonas88) [reply src/content/1.7/modules/sample-modules/grid-and-identifiable-object-form-hooks-usage.md:318]: it's not the right place to put this statement

File: src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md (hunk 11 above)
Comment (tdavidsonas88) [reply src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md]: SOLID practices are good as long as they are for the good purpose. If you are not sure then I'm double not sure @matks !

Comment (tdavidsonas88) [reply src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md]: I think as maintainable is enough and clear is not needed !

Comment (tdavidsonas88) [reply src/content/1.7/modules/sample-modules/extending-sf-form-with-upload-image-field.md]: Supported would be the best !

**Comment by matks** [comment]
Thank you @tdavidsonas88
